content/advent-2017/mphf.md

```diff
@@ -0,0 +1,297 @@
+key in an array, and just walks down the array looking for a match, then
+returns that integer. Obviously this maps each element to a distinct value,
+and is quick to construct. However, both the space required and the execution
+time are not optimal. We can do better.
+
+Let’s start with a very basic implementation. Here’s an example set of keys.
+Let’s pretend they’re commands for some simple network protocol, like NATS or
+Redis. We’ll read 4 bytes from the network and we want to check if we have a
+valid command before dispatching to the appropriate processing loop.
+
+Here are the commands:
+```
+INFO
+CONN
+PUB 
+SUB 
+UNSU
+PING
+PONG
++OK 
+-ERR
+AUTH
+PUSH
+ADD 
+DECR
+SET 
+GET 
+QUIT
+```
+
+I’ve left the spaces after the three letter commands so that each key fits
+into a uint32. In fact, we won't deal with these as strings but we'll turn
+them directly into uint32s.
+
+Here's our first hash function. It takes the uint32 and returns the bottom 4 bits.
+
+```
+func hash1(x uint32) uint32 {
+    return x & 15
+}
+```
+
+And then check for duplicates:
+
+```
+dups := make(map[uint32]bool)
+for _, k := range keys {
+    h := hash1(k)
+    if dups[h] {
+        log.Println("duplicate found")
+    }
+    dups[h] = true
+}
+```
+
+This is very fast, but when we test, half of the keys collide. Changing the
+function to look at the upper 4 bits doesn't work either. The duplicate
+initial letters (PUSH, PUB) and trailing letters (PONG, PING) means we need
+to do more shuffling.
+
+Here’s a second attempt:
+
+```
+const multiplier = 31
+func hash2(x uint32) int {
+   var u32 = (x * multiplier) >> 28
+   return u32 & 15
+}
+```
+
+Here we’ve made two changes. First, we’re multiplying by 31, a nice random
+number that shows up in hash functions. Second, we’re going to extract the
+high bits of the result. High bits of multiplications tend to have a bit more
+entropy than the low bits, another common hash function trick.
+
+This is an improvement, although there are still 6 collisions, down from 8.
+But now we have a framework we can use. Can we find a value for `multiplier`
+that eliminates all collisions? This is easy enough to brute force.
+
+```
+2017/11/29 22:18:25 len(keys)= 16
+2017/11/29 22:18:25 searching for multiplier to fit 16 keys in 16 slots
+m = 715138
+INFO 3
+CONN 6
+PUB  15
+SUB  7
+UNSU 13
+PING 8
+PONG 0
++OK  1
+-ERR 4
+AUTH 2
+PUSH 11
+ADD  14
+DECR 5
+SET  10
+GET  9
+QUIT 12
+```
+
+And indeed, when we set the hash function to use `715138` we
+see that it does indeed map the 16 keys perfectly into 0 to 15.
+
+So in order to check if the bytes we've read are valid, we hash them with our
+perfect hash function and look at the appropriate index in an array. If it's
+the key we're looking for, then we know it's valid. This will be fast because
+the arrays are small and we're just comparing two uint32s.
+
+```
+// in the order listed above
+var values = []uint32{
+	u32s("PONG"),
```

Comment:
For those playing along at home, is there an implementation of `u32s` someplace?

Reply:
Added.